### PR TITLE
Add Weights and Biases Integration

### PR DESCRIPTION
**Pull request by ayulockin**
This PR adds support for [Weights and Biases](https://wandb.ai/site) metric and model checkpointing.

# Usage

I have tested the implementation by training ConvNext Tiny on CIFAR 100 dataset for 10 epochs. To enable logging metrics using W&B use `--enable_wandb true`. To save the model checkpoints as versioned [artifacts](https://docs.wandb.ai/guides/artifacts/model-versioning) use `--wandb_ckpt true` along with `--enable_wandb`. 

```
python main.py --epochs 10 --model convnext_tiny --data_set CIFAR --data_path datasets --num_workers 8 --warmup_epochs 0  --save_ckpt true --output_dir model_ckpt --finetune path/to/model.pth --cutmix 0 --mixup 0 --lr 4e-4 --enable_wandb true --wandb_ckpt true
```

You can also set the name of the W&B project where all the runs will be logged using the `--project` argument (`default='convnext'`).

# Result (Feature Addition)

* Able to easily share the experiments. Here's an **[example W&B run](https://wandb.ai/ayush-thakur/convnext/runs/2aud3vtf)** that shows train and test metrics from fine-tuning a ConvNext tiny on CIFAR 100 for 10 epochs. 
* Ability to check out the exact configuration used to train the model. Configs can thus be easily compared. 
* CPU/GPU metrics - memory allocated, memory utilized, etc. 
* Versioned model checkpoints which are easy to share. 

Note: I achieved a ~88% test top 1 accuracy which is really impressive. Thanks for working on this. 

The screen recording below summarizes the features being added for a few lines of code. 

https://user-images.githubusercontent.com/31141479/150026875-91e4f38d-f465-49b6-9f12-3c8559c2934a.mov

# Note

* I have tested the implementation on a single machine with one GPU and with 2 GPUs and seems to work as expected. You can find the [W&B run](https://wandb.ai/ayush-thakur/convnext/runs/3gkbcdbq) associated with training on 2 K80s.
* If `--enable_wandb true` is not passed the code will work as expected, thus the addition is not enforcing any dependency on W&B.

### CODE REVIEW

**Comment by ayulockin** [comment]
# Additional Feature (note added)

I have also added a feature that allows to interactively visualize model prediction during evaluation. However, I have not added it as part of this PR to make things minimal. The said feature leverages [W&B Tables](https://docs.wandb.ai/guides/data-vis). 

This feature adds a few lines of code in the `evaluate` function defined in `engine.py` as shown below. 

image

The `add_eval_data` and `log_eval_table` methods are belong to the `WandbLogger` class defined in `utils.py`. 

# Usage 

If `--eval true` is used pass `--enable_wandb true` and `--log_eval true` . 

# Result

You can check out the [W&B run page](https://wandb.ai/ayush-thakur/convnext/runs/1mu773wo) where the evaluation table is logged. Note that I have used the fine-tuned weights from the experiment linked above.


https://user-images.githubusercontent.com/31141479/150029459-34bddae0-d75a-436d-bc59-a1fb273b605d.mov

If you think that this would be a useful addition for the community in general, I would love to update this PR or open a new one. :D

**Comment by s9xie** [comment]
Hi Ayush,

Thank you for the PR and additional features. We are particularly interested to see the use of W&B with a CIFAR finetuning example.
I wonder what's the best way to provide some documentation/guidance for people to try the experiment. Is it possible for you to provide a minimal colab notebook for people to get started? We can link it in the main readme.

**Comment by ayulockin** [comment]
Yes, I would be able to provide a minimal colab notebook. 

Do you want me to use my fork of the repo to use W&B logging while fine-tuning on CIFAR? I can later update the colab when the PR gets merged. 

Thank you for your quick response.

**Comment by s9xie** [comment]
> Do you want me to use my fork of the repo to use W&B logging while fine-tuning on CIFAR? I can later update the colab when the PR gets merged.

That sounds good! Thank you!

**Comment by ayulockin** [comment]
Hey @s9xie here's a colab notebook to fine-tune on CIFAR-10 dataset: https://colab.research.google.com/drive/1ijAxGthE9RENJJQRO17v9A7PTd1Tei9F?usp=sharing

Do you want me to update the README.md as part of this PR with the colab link? 

> I have also added a feature that allows to interactively visualize model prediction during evaluation. However, I have not added it as part of this PR to make things minimal. The said feature leverages W&B Tables.

What do you think about this feature?

**Comment by s9xie** [comment]
> https://colab.research.google.com/drive/1ijAxGthE9RENJJQRO17v9A7PTd1Tei9F?usp=sharing

Thanks for the notebook. It looks good! The last image in the conclusion block seems broken on my end though.

> Do you want me to update the README.md as part of this PR with the colab link?
I think it is not necessary, I can add it after the merge.

> What do you think about this feature?

Let's keep it minimal for now. This is a great feature but I think for people who want to use it, they can find relevant information on W&B documentation. 

I will go ahead and merge if there's no more update. And I assume you can update the notebook and remove the fork note?

**Comment by ayulockin** [comment]
Thanks for your response. 

> The last image in the conclusion block seems broken on my end though.

Are you using the Brave browser to view the colab? If so can you view it using Chrome, Safari or Mozilla and let me know if it's broken?

> Let's keep it minimal for now.

Totally.

> I will go ahead and merge if there's no more update.

There is one final tweak I want to do before it get merged. Currently the logging is after every epoch and it takes sometime for the first epoch to end. Somone new to W&B might get confused with this behavior. The update is to log the metrics after every `print_frequency` aswell. It should be a minor change. 

>  I assume you can update the notebook and remove the fork note.

Will do.

**Comment by s9xie** [comment]
I'm using chrome - this is what I see on my end.
image

**Comment by ayulockin** [comment]
I will fix it. Thanks for letting me know.

**Comment by ayulockin** [comment]
Hey, @s9xie I am ready on my side for the merge. Please let me know if you need any changes. :D

**Comment by s9xie** [comment]
@ayulockin LGTM. Thanks for the contribution!

**Comment by ayulockin** [comment]
Will do and let you know. :D
